File: algorithms/111.py
# https://leetcode.com/problems/minimum-depth-of-binary-tree


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0

        left = self.minDepth(root.left)
        right = self.minDepth(root.right)
        return left + right + 1 if (left == 0 or right == 0) else min(left, right) + 1

File: algorithms/111.swift
// https://leetcode.com/problems/minimum-depth-of-binary-tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func minDepth(root: TreeNode?) -> Int {
        func dfs(node: TreeNode?, _ count: Int) -> Int {
            guard let n = node else { return count }
            let count = count + 1

            if n.left != nil && n.right != nil {
                return min(dfs(n.left, count), dfs(n.right, count))
            }
            else {
                return max(dfs(n.left, count), dfs(n.right, count))
            }
        }
        return dfs(root, 0)
    }
}
